### PR TITLE
Add base abstractions for tags storing

### PR DESCRIPTION
**Pull request by borovskyav**

### CODE REVIEW

File: Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        
+        [Test]
+        public void ToString_should_return_empty_string_then_tag_collection_is_empty()
+        {
+            new TagCollection().ToString().Should().Be("");
+        }
+
+        [Test] // todo should think about edge cases
+        public void TryParse_should_return_empty_tag_collection_on_bad_string()
```
Comment (iloktionov) [inline Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs]: А почему так? Из TryParse же обычно принято возвращать false, если input некорректного формата!

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+namespace Vostok.ServiceDiscovery.Abstractions.Models
+{
+    [PublicAPI]
+    public class TagCollection : Dictionary<string, string>
+    {
+        private const string DefaultValue = "true";
+        private const string TagsSeparator = "|";
+        private const string TagsKeyValueSeparator = "=";
+        
+        private static readonly DictionaryComparer<string, string> DictionaryComparer = new DictionaryComparer<string, string>(StringComparer.OrdinalIgnoreCase);
```
Comment (iloktionov) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: Обрати внимание, что comparer, переданный тут в DictionaryComparer — это valueComparer. У нас точно значения тегов case-insensitive?

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                        var key = input.Substring(iterator, nextKeyValueSeparator - iterator);
+                        var value = input.Substring(nextKeyValueSeparator + 1, nextTagSeparator - nextKeyValueSeparator - 1);
+                        tagCollection[key] = value;
+                    }
+                    iterator = nextTagSeparator + 1;
+                }
+                
+                return true;
+            }
+            catch
```
Comment (iloktionov) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: Точно хотим съесть тут любые эксепшны? Кажется, что в методе нет ни одной уважительной причины для них, кроме логического бага в коде, и лучше уж его заметить по взрыву, чем гадать, почему не парсится..

File: Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
 
 namespace Vostok.ServiceDiscovery.Abstractions.Models
 {
+    [PublicAPI]
     public class ReplicaInfo : IReplicaInfo
     {
         private readonly Dictionary<string, string> properties;
 
-        public ReplicaInfo([NotNull] string environment, [NotNull] string application, [NotNull] string replica, [CanBeNull] Dictionary<string, string> properties = null)
+        public ReplicaInfo([NotNull] string environment, [NotNull] string application, [NotNull] string replica, [CanBeNull] Dictionary<string, string> properties = null, [CanBeNull] TagCollection tags = null)
```
Comment (iloktionov) [inline Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: Боюсь, тут лучше не трогать конструктор и просто добавить сеттер для свойства Tags.

Почему: в рантайме изменится сигнатура метода, и это ломающее изменение. Представим себе, что приложение тянет либы из нугета, и подтащило новейшие абстракции, но при этом использует старую версию модуля с маячком. Старая либа маячка попытается вызвать конструктор без TagCollection и поймает MissingMethodException несмотря на то, что всё бы работало, если заново перекомпилировать все зависимости с новейшими версиями.

File: Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs (hunk 1 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs]: Для меня пока это спорный вопрос - у меня есть 3 варианта:
1. Тупо игнорить такие кейсы, про них забывать и из какого-нибудь `tag1|=tag2|tag3=|tag4=value` делать коллекцию с одним тегом.
2. Честно кидать false если есть хотябы одна нода некоректного формата.
3. Кидать false только если все ноды некоректного формата.

И я пока не знаю как здесь правильно зафиксироваться.
Но а вообще я пока еще даже не понял, может ли быть tagKey или tagValue пустой строкой.
Сейчас как обычно отвлекся на другую полуавтоматическую писанину чтобы подумать в фоне.

File: Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs (hunk 4 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: Да косяк, мне еще предстоит научиться думать о таких вещах, поправлю конечно.

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs (hunk 2 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: Да, я намеренно сделал значения case-insensitive, как это сделано с ключами, не увидел в этом большой проблемы, кажется здесь как договоримся.

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs (hunk 3 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: Да, кажется переборщил, согласен. Про true/false я еще хочу отдельно подумать (это к слову про коммент в тестах), я пока все же не понимаю как правильно поступать: молча проглатывать кривые теги или выкидывать false

File: Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs (hunk 4 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: Слушай, я не знаю почему, но мне почему-то очень не нравится делать сеттер, нам что-то мешает сделать еще один конструктор? Кажется это конечно больше microsoft-way пилить вот такие конструкции, но все же
```
public ReplicaInfo([NotNull] string environment, [NotNull] string application, [NotNull] string replica, [CanBeNull] Dictionary<string, string> properties, TagCollection tags)
{
    if (string.IsNullOrWhiteSpace(environment))
        throw new ArgumentOutOfRangeException(nameof(environment), environment);
    if (string.IsNullOrWhiteSpace(application))
        throw new ArgumentOutOfRangeException(nameof(application), application);
    if (string.IsNullOrWhiteSpace(replica))
        throw new ArgumentOutOfRangeException(nameof(replica), replica);

    this.properties = properties ?? new Dictionary<string, string>();
    Environment = environment;
    Application = application;
    Replica = replica;
}

public ReplicaInfo([NotNull] string environment, [NotNull] string application, [NotNull] string replica, [CanBeNull] Dictionary<string, string> properties)
    : this(environment, application, replica, properties, null)
{
}

public ReplicaInfo([NotNull] string environment, [NotNull] string application, [NotNull] string replica)
    : this(environment, application, replica, null, null)
{
}
```

Comment (kungasc) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: 2 конструктора очень copy-paste выглядят
ещё один новый параметр и их может 4 стать

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs (hunk 2 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: Кажется в case-(in)sensitive споре у меня побеждает оставить значения case-sensitive с таким соображением: всё круто, мы научились делать внутренние операции case-insensitive, но никак не объяснили это пользователю, он же может взять просто `tagCollection[key] == value` сделать, я не понимаю насколько это проблема, но правда боюсь что пользователь выстрелит себе этим в ногу, коллекции вроде на Equals возвращают true, а значения не совпадают.

Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: done

File: Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs (hunk 1 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs]: Фиксирую логику:
Валидный ключ+значение - это пара строк, разделенная знаком `=`. Пример `key=value`, `key=`, `value=`, `key=val=ue`.
Из TryParse возвращаем `true` и заполненный `TagCollection` когда все пары, разделенные знаком `|` валидны.
Если хотябы одна пара не валидна - возвращаем `false` и `TagCollection=null`
Так же невалидной считается строка, в которой лежит 2 пары с одинаковым ключём, на такое тоже возвращем `false`

File: Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs (hunk 4 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: Для пользователя я в копипасте никакой проблемы не вижу. Для нас - мы не так часто меняем интерфейс конструктора чтобы эти штуки плодились с большой скоростью

File: Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs (hunk 1 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs]: done.
Меня всё-же мучает вопрос со знаком равно и с кейсом `key=val=ue`, как вам идея заменить его на знак [═](https://unicode-table.com/ru/2550/)?
На самом деле не понимаю что лучше сделать - экранировать или обойтись вот таким знаком.

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs (hunk 3 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: done

File: Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs (hunk 1 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions.Tests/TagCollection_Tests.cs]: В общем сделал, теперь `key=val=ue` не пройдет

File: Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs (hunk 4 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: сделал сеттер

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                return false;
+            
+            tagPropertyKey = new TagPropertyKey(split[1], split[2]);
+            return true;
+        }
+
+        public override string ToString()
+            => TagsParameterPrefix + TagsParameterValuesSeparator + ReplicaName + TagsParameterValuesSeparator + TagKind;
+        
+        private const string TagsParameterPrefix = "Tags";
```
Comment (iloktionov) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]: Кажется, если применить code style, это должно будет уехать вверх класса.

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        [NotNull]
+        public string ReplicaName { get; }
+        
+        /// <summary>
+        /// 
+        /// </summary>
+        [NotNull]
+        public string TagKind { get; }
+
+        public static bool TryParse([NotNull] string input, out TagPropertyKey tagPropertyKey)
```
Comment (iloktionov) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]: Судя по коду в extensions, этот метод будет вызываться на всех свойствах. То есть наделает кучу лишних `Split()` на свойствах, которые вообще не о тегах. Может, поэффективнее отсеем их (к тому же, тут это тривиально)?

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            tagPropertyKey = null;
+            var split = input.Split(TagsParameterValuesSeparator.ToCharArray());
```
Comment (iloktionov) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]: Зачем каждый раз ToCharArray? Лишний мусор.

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs (hunk 5 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]: отреформатил, заодно отреформатил весь csproj, получилось много файлов, решил не коммитить, может в отдельном PR отреформатим, или не так критично остальное? (там в основном отступы всякие)

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs (hunk 6 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]: Убрал Split, буду делать Substring, как раньше.

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs (hunk 7 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]: done

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using Vostok.Commons.Helpers.Comparers;
+
+namespace Vostok.ServiceDiscovery.Abstractions.Models
+{
+    [PublicAPI]
+    public class TagCollection : Dictionary<string, string>
```
Comment (kungasc) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]:
```suggestion
    public class TagsCollection : Dictionary<string, string>

```

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            : base(StringComparer.OrdinalIgnoreCase)
+        {
+        }
+
+        public TagCollection(IDictionary<string, string> tagsCollection)
+            : base(tagsCollection, StringComparer.OrdinalIgnoreCase)
+        {
+        }
+
+        public static bool TryParse([CanBeNull] string input, out TagCollection tagCollection)
```
Comment (kungasc) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs:31]: Почему не просто string.Split('|')? Да и равно, которое не равно выглядит костылём, потом кто-нибудь посмотрит и напишет обычное.
https://github.com/vostok/servicediscovery/blob/master/Vostok.ServiceDiscovery/Serializers/ReplicaNodeDataSerializer.cs#L32

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
+using System;
+using JetBrains.Annotations;
+
+namespace Vostok.ServiceDiscovery.Abstractions.Models
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    [PublicAPI]
+    public class TagPropertyKey
```
Comment (kungasc) [inline Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]:
```suggestion
    public class TagsPropertyKey
```

File: Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs [hunk 11]
```diff
@@ -29,6 +30,8 @@ public ReplicaInfo([NotNull] string environment, [NotNull] string application, [
 
         public string Replica { get; }
 
+        public TagCollection Tags { get; set; }
```
Comment (kungasc) [inline Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: А это зачем?

File: Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs (hunk 10 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagPropertyKey.cs]: done

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs (hunk 8 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs]: Кажется решили не делать

File: Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs (hunk 9 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs:31]: > Да и равно, которое не равно выглядит костылём, потом кто-нибудь посмотрит и напишет обычное.

У меня было 2 варианта решения [экранировать или использовать какой-нибудь редко используемый знак](https://github.com/vostok/servicediscovery.abstractions/pull/2#discussion_r511486816), решил выбрать редкоиспользуемый знак, как вариант - заменить на что-нибдуь более заметное.

Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs:31]: >Почему не просто string.Split('|')?

У нас с @iloktionov было обсуждение по поводу этого, но я не нашел где. Обсуждение сводилось к тому что здесь есть промежуточный Split (как раз вот этот ` string.Split('|')`) генерит много мусора, который никак не используется в дальшейшем, поэтому я переписал на решение в лоб, вот что давали замеры бенчмарком:

```
| Method        |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
|-------------- |----------:|----------:|----------:|-------:|----------:|
|     Split     | 11.025 us | 0.2357 us | 0.6838 us | 1.5564 |   6.43 KB |
|     Iterator  |  3.615 us | 0.0672 us | 0.1403 us | 0.5188 |   2.14 KB |
```

Так как TryParse в некоторых сценариях будет использоваться очень часто, кажется это заметный выигрыш.

Comment (kungasc) [reply Vostok.ServiceDiscovery.Abstractions/Models/TagCollection.cs:31]: можно ещё хранить бинарным мясом, как веса и блеклист :)

File: Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs (hunk 11 above)
Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: Я решал конкретную задачу: мне нужно было научиться красиво в `IReplicaInfoBuilder`e передавать теги. Я думал над двумя решениями (оба плохие по моему мнению):
1. Расширить класс `ReplicaInfo` не трогая при этом `IReplicaInfo`.
2. Выставить в `IReplicaInfoBuilder` костыль в виде метода `GetTags` и научиться дергать методы из билдера в internal конструкторах.

Я склонился к первому варинту в первую очередь потому что не увидел в этом больших проблем:
- ServiceBeacon везде где только можно использует `ReplicaInfo`, а не интерфейс.
- Те места, которые возвращают `IReplicaInfo` используют именно тип интерфейса, а не класса.

При этом мне не понятно, насколько сильно мы хотим гавнякать в `IReplicaInfoBuilder`е

Comment (kungasc) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: Так кажется, что методы ServiceDiscoveryManager GetReplicaAsync и TryCreatePermanentReplicaAsync будут как-то учитывать это поле.
Поэтому я за второй вариант, раз уж теги на реплику у нас пока только в биконе используются в контексте реплики.

Comment (borovskyav) [reply Vostok.ServiceDiscovery.Abstractions/Models/ReplicaInfo.cs]: Выпилил, в https://github.com/vostok/servicediscovery/pull/7 в `ReplicaInfoBuilder.Build` теперь возвращаю новую абстрацию
